### PR TITLE
Make the storage-backend flag mandatory

### LINKED ISSUE #45457: storage-backend should be consistent across all servers
apiserver has storage-backend default to etcd2.  In 1.6, we're hoping to use UAS for additional API functionality, and for new servers the default has been changed to etcd3.

This is bad UX; it ends up where we have to know the default version of etcd in place at the time when the server project was first created (etcd2, 3, 4, 5...).  Not passing any argument will result in a different backend assumption for apiserver than for a UAS.

We should either make etcd2 the default everywhere, or simply make specifying storage-backend mandatory (we could deprecate defaulting of the storage-backend, in our normal process).

### PR DESCRIPTION
**Pull request by justinsb**
The default value has changed in the past, which was a violation of our
deprecation policy, but also shows that we can't have a default while
maintaining both sensible defaults and a deprecation policy.

As such, make the flag mandatory: the installer knows best which version
of etcd or other storage backend they are choosing to install.

Fix #45457

```release-note
The --storage-backend flag must now be passed to apiserver.  It is not possible to both have a sensible default and respect the deprecation policy, so the storage-backend must be specified.
```

### CODE REVIEW

**Comment by justinsb** [comment]
Fully expect this to break at least one e2e - it'll be valuable signal to know how many though...

File: staging/src/k8s.io/apiserver/pkg/server/options/etcd.go [hunk 1]
```diff
@@ -88,7 +88,7 @@ func (s *EtcdOptions) AddFlags(fs *pflag.FlagSet) {
 		"Enable watch caching in the apiserver")
 
 	fs.StringVar(&s.StorageConfig.Type, "storage-backend", s.StorageConfig.Type,
-		"The storage backend for persistence. Options: 'etcd3' (default), 'etcd2'.")
+		"The storage backend for persistence. Options: 'etcd2', 'etcd3'.")
```
Comment (xiangpengzhao) [inline staging/src/k8s.io/apiserver/pkg/server/options/etcd.go]: Do we need to validate this flag at some place?

Comment (xiangpengzhao) [inline staging/src/k8s.io/apiserver/pkg/server/options/etcd.go]: Then users can find out there config mistake as early as possible.

Comment (justinsb) [inline staging/src/k8s.io/apiserver/pkg/server/options/etcd.go]: Validation happens in `Create(c storagebackend.Config)`

Comment (xiangpengzhao) [inline staging/src/k8s.io/apiserver/pkg/server/options/etcd.go]: yeah I saw Create. I meant is there another place ahead of it where we can validate the flag? I might understood incorrectly...

Comment (justinsb) [inline staging/src/k8s.io/apiserver/pkg/server/options/etcd.go]: That would be a separate PR

Comment (xiangpengzhao) [inline staging/src/k8s.io/apiserver/pkg/server/options/etcd.go]: ah, I see there is `Validate` in L61 of this file. How about validate the flag there?

Comment (justinsb) [inline staging/src/k8s.io/apiserver/pkg/server/options/etcd.go]: That would be a separate PR

Comment (xiangpengzhao) [inline staging/src/k8s.io/apiserver/pkg/server/options/etcd.go]: Agreed:)

Comment (justinsb) [inline staging/src/k8s.io/apiserver/pkg/server/options/etcd.go]: :-)

**Comment by smarterclayton** [comment]
So when someone upgrades, they must update all their config or their cluster is broken, right?

**Comment by justinsb** [comment]
Sadly yes, changing the flag default in 1.6 did that.

This does force another breakage on users  if they aren't specifying the flag, but by now everyone should have already been burned by 1.6.  This breakage has a nice error message that is easy to understand IMO (`storage type must be set (use --storage-backend flag)` currently).

The value of doing it now is that we don't break _again_ when etcd4 comes out.  And the installer (human or automated) knows full well which storage-backend they are installing - hopefully!

cf discussion on the linked issue.

**Comment by smarterclayton** [comment]
I didn't see any discussion about "we broke them in 1.6 so we'll break them again"

**Comment by justinsb** [comment]
Ah, you're right - discussion also happened on the linked-linked-PR e.g. https://github.com/kubernetes/kubernetes/pull/46167#issuecomment-303735500

We can have the discussion here though, as a clean discussion.  What do you propose instead?

**Comment by smarterclayton** [comment]
Never change the default again (which you're right, was a mistake).  Make
the field mandatory in the new api server config API objects we create
(component config + successors).  Not specified means etcd3 (as it does in
1.6) forever.

**Comment by justinsb** [comment]
I think we are agreed that we cannot change the default again when etcd4 comes out.

For today's behaviour though:

I guess your vote would be to "keep etcd3 forever", which is reasonable, though IMO bad UX as we eventually deprecate etcd3.

My vote would be to "make it mandatory now", because we're about to have a large number of UAS servers and so we can't change this later.

We currently really only have kube-apiserver (?);  anyone that is on on 1.5 today will then experience one breakage with a nicer error message than they otherwise would, anyone that is currently on 1.6 should be specifying the flag by now.

**Comment by justinsb** [comment]
I guess though that we could make the flag mandatory when we remove etcd3 support?  That isn't horrible, assuming we remove etcd3 promptly once we have a new recommendation.

**Comment by smarterclayton** [comment]
Yeah, that would be a good time for it.

**Comment by liggitt** [comment]
> I guess though that we could make the flag mandatory when we remove etcd3 support?

That's what I would expect.

**Comment by justinsb** [comment]
Sounds like we want to solve this through policy.  Closing.
